From 9e2f6d6fab1926f21d423ab8d5c9909a13ef3971 Mon Sep 17 00:00:00 2001
From: "Bernhard R. Link" <brlink@debian.org>
Date: Sat, 15 Jan 2005 12:00:00 +0100
Subject: add --with-menu

This patch adds an --with-menu that allows to specify a menu command
that will be available with :menu and with "C-t ."

* configure.in: add --with-menu to specify a menu program
* src/action.c: add menu alias and binding if menu specified
* src/main.c:   tell menu command if menu specified
---
 configure.in   |    9 +++++++++
 src/actions.c  |    7 +++++++
 src/main.c     |   45 ++++++++++++++++++++++++++++++++++-----------
 src/messages.h |    1 +
 4 files changed, 51 insertions(+), 11 deletions(-)

diff --git a/configure.in b/configure.in
index 2d37a81..68b6fa9 100644
--- a/configure.in
+++ b/configure.in
@@ -72,6 +72,15 @@ fi
 AC_SUBST(XFT_CFLAGS)
 AC_SUBST(XFT_LIBS)
 
+AC_ARG_WITH(menu, [ --with-menu=PROG        set a external menu program to be advertised ],
+menu_prog="$withval", menu_prog="")
+
+if test "$menu_prog" != "no" && ! test -z "$menu_prog" ; then
+AC_DEFINE_UNQUOTED(MENU_PROG, "$menu_prog", external menu program to advertise)
+AC_MSG_CHECKING(external menu program)
+AC_MSG_RESULT($menu_prog)
+fi
+
 dnl Checks for programs.
 AC_CHECK_TOOL(CC, gcc)
 AC_PROG_CC
diff --git a/src/actions.c b/src/actions.c
index d1a46dd..972ba50 100644
--- a/src/actions.c
+++ b/src/actions.c
@@ -725,6 +725,9 @@ initialize_default_keybindings (void)
   add_keybinding (prefix_key.sym, prefix_key.state, "other", map);
   add_keybinding (prefix_key.sym, 0, "meta", map);
   add_keybinding (XK_g, RP_CONTROL_MASK, "abort", map);
+#ifdef MENU_PROG
+  add_keybinding (XK_period, 0, "menu", map);
+#endif
   add_keybinding (XK_0, 0, "select 0", map);
   add_keybinding (XK_1, 0, "select 1", map);
   add_keybinding (XK_2, 0, "select 2", map);
@@ -811,6 +814,10 @@ initialize_default_keybindings (void)
   add_alias ("unbind", "undefinekey " ROOT_KEYMAP);
   add_alias ("bind", "definekey " ROOT_KEYMAP);
   add_alias ("split", "vsplit");
+
+#ifdef MENU_PROG
+  add_alias ("menu", "exec " MENU_PROG );
+#endif
 }
 
 cmdret *
diff --git a/src/main.c b/src/main.c
index c7ad72f..69c868c 100644
--- a/src/main.c
+++ b/src/main.c
@@ -467,7 +467,13 @@ show_welcome_message (void)
 {
   rp_action *help_action;
   char *prefix, *help;
+  const char *help_show;
   rp_keymap *map;
+#ifdef MENU_PROG
+  rp_action *menu_action;
+  char *menu;
+  const char *menu_show;
+#endif
 
   prefix = keysym_to_string (prefix_key.sym, prefix_key.state);
 
@@ -476,27 +482,44 @@ show_welcome_message (void)
   /* Find the help key binding. */
   help_action = find_keybinding_by_action ("help " ROOT_KEYMAP, map);
   if (help_action)
-    help = keysym_to_string (help_action->key, help_action->state);
-  else
-    help = NULL;
-
-
-  if (help)
     {
+      help = keysym_to_string (help_action->key, help_action->state);
       /* A little kludge to use ? instead of `question' for the help
          key. */
       if (!strcmp (help, "question"))
-        marked_message_printf (0, 0, MESSAGE_WELCOME, prefix, "?");
+        help_show = "?";
       else
-        marked_message_printf (0, 0, MESSAGE_WELCOME, prefix, help);
-
-      free (help);
+	help_show = help;
     }
   else
     {
-      marked_message_printf (0, 0, MESSAGE_WELCOME, prefix, ":help");
+      help = NULL;
+      help_show = ":help";
+    }
+#ifdef MENU_PROG
+  /* Find the menu key binding. */
+  menu_action = find_keybinding_by_action ("menu", map);
+  if (menu_action)
+    {
+      menu = keysym_to_string (menu_action->key, menu_action->state);
+      if (!strcmp (menu, "period"))
+	menu_show = ".";
+      else
+	menu_show = menu;
+    }
+  else
+    {
+      menu = NULL;
+      menu_show = ":menu";
     }
 
+  marked_message_printf (0, 0, MESSAGE_WELCOME_MENU, prefix, help_show,
+			 prefix, menu_show);
+  free(menu);
+#else
+  marked_message_printf (0, 0, MESSAGE_WELCOME, prefix, help_show);
+#endif
+  free(help);
   free (prefix);
 }
 
diff --git a/src/messages.h b/src/messages.h
index 47f2563..ce5ed90 100644
--- a/src/messages.h
+++ b/src/messages.h
@@ -50,6 +50,7 @@
 #define MESSAGE_PROMPT_VAR_VALUE  "Value: "
 
 #define MESSAGE_WELCOME "Welcome to ratpoison! Hit `%s %s' for help."
+#define MESSAGE_WELCOME_MENU 	"Welcome to ratpoison! Hit `%s %s' for help. `%s %s' for menu."
 
 #define EMPTY_FRAME_MESSAGE "Current Frame"
 
